src/parse.rs

```diff
@@ -0,0 +1,159 @@
+        Err(e) => return Err(FenError::ParseErr(e)),
+        Ok((_, (parsed_board, parsed_active_player, parsed_castling_rights, parsed_en_passant_target, half_turn_clock, full_turn_clock))) => {
+            let mut piece_position: [[char;8];8] = [['.';8];8];
+            for (num, row) in parsed_board.into_iter().enumerate(){
+                let mut result_row: Vec<char> = vec![];
+                for c in row {
+                    match c {
+                        '1'..='8' => {
+                            let x = (c as u8) - b'0';
+                            result_row.append(&mut vec!['.'; x.into()]);
+                        }
+                        x => result_row.push(x)
+                    };
+                }
+                if result_row.len() != 8 {
+                    return Err(FenError::InvalidRow(result_row));
+                }
+                else {
+                    piece_position[num] = result_row[0..8].try_into().unwrap();
+                }
+            }
+            let active_player = match parsed_active_player {
+                'b' => ActivePlayer::Black,
+                'w' => ActivePlayer::White,
+                x => return Err(FenError::InvalidActivePlayer(x))
+            };
+            let black_kingside = parsed_castling_rights.contains('k');
+            let black_queenside = parsed_castling_rights.contains('q');
+            let white_kingside = parsed_castling_rights.contains('K');
+            let white_queenside = parsed_castling_rights.contains('Q');
+            let castling_rights = CastlingRights{black_kingside, black_queenside, white_kingside, white_queenside};
+            let en_passant_target = match parsed_en_passant_target {
+                "-" => None,
+                x => {
+                    let rank = match Rank::try_from(x.as_bytes()[0] - b'a'){
+                        Ok(rank) => rank,
+                        Err(x) => return Err(FenError::InvalidRank(x))
+                    };
+                    let file = match File::try_from(x.as_bytes()[0] - b'a'){
+                        Ok(file) => file,
+                        Err(x) => return Err(FenError::InvalidFile(x))
+                    };
+                    Some((rank, file))
+                }
+            };
+            Ok(ParsedGameState{piece_position, active_player, castling_rights, en_passant_target, half_turn_clock, full_turn_clock })
+        }
+    }
+}
+
+fn fen_board(input: &str) -> IResult<&str, Vec<Vec<char>>>{
+    count(terminated(many1(one_of("rnbqkpRNBQKP12345678")), opt(char('/'))), 8)(input)
+}
+
+fn fen_active_player(input: &str) -> IResult<&str, char> {
+    one_of("wb")(input)
+}
+
+fn fen_castling(input: &str) -> IResult<&str, &str> {
+    alt((tag("-"),recognize(many1(one_of("KQkq")))))(input)
+}
+
+fn fen_en_passant_target(input: &str) -> IResult<&str, &str> {
+    alt((tag("-"), recognize(pair(one_of("abcdefgh"), one_of("12345678")))))(input)
+}
```

Comment:
A minor nitpick, but I think defining these after they're used can make things a little tricky. Feel free to ignore though